core/chain_makers.go

```diff
@@ -340,3 +366,98 @@ func (cr *fakeChainReader) GetGasUsedInChain(block *types.Block, length int) int
 func (cr *fakeChainReader) GetExternalBlocks(header *types.Header) ([]*types.ExternalBlock, error) {
 	return nil, nil
 }
+
+// GenerateNetwork creates a network of n blocks in c contexts. The first block's
+// parent will be the provided parent. db is used to store intermediate states
+// and should contain the parent's state trie.
+
+// configs should contain the respective config settings being tested. This must
+// include a valid slice of Prime, Region, and Zone mining locations. This is
+// necessary for testing interchain linkages through coincident blocks as well
+// as transactions flowing through them. The array of ints in c represent
+// the intended order of contexts to be mined.
+
+// The generator function is called with a new block generator for every
+// block. Any transactions and uncles added to the generator become part of the
+// block. If gen is nil, the blocks will be empty and their coinbase will
+// be the zero address.
+
+// Blocks created by GenerateNetwork do not contain valid proof of work values.
+// Inserting them into BlockChain requires use of a TestPoW in order to test and
+// verify the correct operation of the Proof-of-Work algorithm.
+func GenerateNetwork(primeConfig *params.ChainConfig, regionConfig params.ChainConfig, zoneConfig params.ChainConfig, parent *types.Block, engine consensus.Engine, db ethdb.Database, n int, orders []int, gen func(int, *BlockGen)) ([]*types.Block, []types.Receipts) {
+	// the first config must be the Prime
+	if primeConfig != params.MainnetPrimeChainConfig {
+		print("must have MainnetPrimeChainConfig as first config")
+	}
+	// check second config for Region
+	if !(regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[0].ChainID) == 0 || regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[1].ChainID) == 0 || regionConfig.ChainID.Cmp(params.MainnetRegionChainConfigs[2].ChainID) == 0) {
+		print("must have a MainnetRegionChainConfig as second config")
+	}
+	// check third config for Zone in Region
+	if !(zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][0].ChainID) == 0 || zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][1].ChainID) == 0 || zoneConfig.ChainID.Cmp(params.MainnetZoneChainConfigs[regionConfig.Location[0]-1][2].ChainID) == 0) {
+		print("must have a MainnetZoneConfig as third config")
+	}
```

Comment:
IIUC it is possible to select a region & zone which do not fall in the same slice. Is that an error? e.g. if I pass in configs for cypress region, but paxos1 zone.

Reply:
Made the change to infer slice from zone config based on your suggestion.